### PR TITLE
[dynamo][guards] Support cloning of Guard Manager

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140251
* #140250
* __->__ #140223

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/csrc/dynamo/guards.cpp [hunk 1]
```diff
@@ -1647,6 +1682,10 @@ class GuardManager {
     }
   }
 
+  // For cloning
+  GuardManager(RootGuardManager* root, std::string source, bool is_dict)
+      : _root(root), _source(source), _is_dict(is_dict) {}
```
Comment (Skylion007) [inline torch/csrc/dynamo/guards.cpp:1687]:
```suggestion
      : _root(root), _source(std::move(source)), _is_dict(is_dict) {}
```

File: torch/csrc/dynamo/guards.cpp [hunk 2]
```diff
@@ -1663,6 +1702,34 @@ class GuardManager {
...
+          accessor->clone(cloned_root, clone_filter_fn)));
+    }
+  }
+
+  virtual GuardManager* clone(
+      RootGuardManager* cloned_root,
+      py::function clone_filter_fn) {
+    // Use clone_filter_fn
+    GuardManager* cloned_mgr = new GuardManager(cloned_root, _source, _is_dict);
+    clone_common(cloned_root, cloned_mgr, clone_filter_fn);
```
Comment (Skylion007) [inline torch/csrc/dynamo/guards.cpp:1729]:
```suggestion
    clone_common(cloned_root, cloned_mgr, std::move(clone_filter_fn));
```

File: torch/csrc/dynamo/guards.cpp [hunk 3]
```diff
@@ -1939,6 +2006,11 @@ class RootGuardManager : public GuardManager {
     _relational_guard_resetters.emplace_back(std::move(relational_guard));
   }
 
+  void add_relational_guard_resetter_without_move(
+      std::shared_ptr<RelationalGuard> relational_guard) {
+    _relational_guard_resetters.emplace_back(relational_guard);
```
Comment (Skylion007) [inline torch/csrc/dynamo/guards.cpp:2011]:
```suggestion
    _relational_guard_resetters.emplace_back(std::move(relational_guard));
```

Comment (Skylion007) [reply torch/csrc/dynamo/guards.cpp:2011]: Surprised this isn't covered by clang-tidy. A lot of things should probably moved by value around here.

File: torch/csrc/dynamo/guards.cpp [hunk 4]
```diff
@@ -2541,8 +2685,45 @@ class DictSubclassGuardManager : public DictGuardManager {
...
+  DictSubclassGuardManager(
+      RootGuardManager* cloned_root,
+      std::string source,
+      Py_ssize_t size,
+      PyTypeObject* _expected_type,
+      bool is_exact_dict_type,
+      std::vector<Py_ssize_t> indices)
+      : DictGuardManager(
+            cloned_root,
+            source,
```
Comment (Skylion007) [inline torch/csrc/dynamo/guards.cpp:2699]:
```suggestion
            std::move(source),
```

File: torch/csrc/dynamo/guards.cpp [hunk 5]
```diff
@@ -2541,8 +2685,45 @@ class DictSubclassGuardManager : public DictGuardManager {
...
+            size,
+            _expected_type,
+            is_exact_dict_type,
+            indices) {}
```
Comment (Skylion007) [inline torch/csrc/dynamo/guards.cpp:2703]:
```suggestion
            std::move(indices)) {}
```

File: torch/csrc/dynamo/guards.cpp [hunk 6]
```diff
@@ -2354,6 +2438,66 @@ class DictGuardManager : public GuardManager {
     return _is_exact_dict_type;
   }
 
+ public: // cloning functions
+  DictGuardManager(
+      RootGuardManager* cloned_root,
+      std::string source,
```
Comment (Skylion007) [inline torch/csrc/dynamo/guards.cpp:2444]:
```suggestion
      std::string std::move(source),
```

**Review by Skylion007 — COMMENTED** [review]
Way to many things should be moved in the newly added code. I went through a few of them, but you should really  use clang-tidy to find all the args that need to moved though the calls.

File: torch/csrc/dynamo/guards.cpp [hunk 7]
```diff
@@ -2541,8 +2685,45 @@ class DictSubclassGuardManager : public DictGuardManager {
...
+    RootGuardManager* cloned_root,
+    py::function clone_filter_fn) {
+  return from->clone(cloned_root, clone_filter_fn);
+}
+
+void add_relational_guard_resetter_to_cloned_root(
+    RootGuardManager* root,
+    std::shared_ptr<RelationalGuard> guard) {
+  // Don't std::move. Shared the pointers between root and cloned_root.
+  root->add_relational_guard_resetter_without_move(guard);
```
Comment (Skylion007) [inline torch/csrc/dynamo/guards.cpp:2724]:
```suggestion
  root->add_relational_guard_resetter_without_move(std::move(guard));
```

**Comment by anijain2305** [comment]
> Way to many things should be moved in the newly added code. I went through a few of them, but you should really use clang-tidy to find all the args that need to moved though the calls.

Thanks for pointing out. This PR will be WIP for a few weeks (trying to build up to a new functionality in follow up PRs). I will fix the std::move problems before the final review.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**